sail/g6/Player.java

```diff
@@ -38,66 +36,59 @@ public Point chooseStartingLocation(Point wind_direction, Long seed, int t) {
         this.targets = targets;
         this.id = id;
+        isVisited = new int[targets.size()][group_locations.size()];
     }
 
     @Override
     public Point move(List<Point> group_locations, int id, double dt, long time_remaining_ms) {
 
-        // Check if we've reached the target.
         if (this.path.size() != 0 && visited_set.get(id).contains(nextTarget)) {
             prevLoc.add(0, path.get(0));
             path.remove(0);
         }
         else if (this.path.size() != 0) {
-            // System.out.println("Moving to set target.");
-
-            // Check if we crossed the intermediate point.
-            // This won't be a target point because that is already checked for in the
-            // first "if".
             if (checkCrossedInterpolation(path.get(0), prevLoc.get(0), group_locations.get(id))) {
                 prevLoc.add(0, path.get(0));
                 path.remove(0);
-                // We are guaranteed that there is another point in "path".
-            }
-
-            if (path.size() == 0) {
-                // We should never come here.
-                // If we do, then it means that the first "if" that checks if
-                // we crossed a target is buggy.
-                System.out.println("Whaaaaaa");
             }
-
-            // Continue moving to the intermediate point or move to the target point.
             return moveInPath(group_locations.get(id), path.get(0));
         }
 
         if(visited_set != null && visited_set.get(id).size() == targets.size()) {
-            // This is if we have finished visiting all targets.
             nextTarget = targets.size();
             return findPathAndMove(group_locations.get(id), initial);
         }
         else {
-            // System.out.println("Selecting next target.");
-
-            // Here is the logic to decide which target to head to next.
-            // Pick the closest target.
+            for (int i = 0; i < targets.size(); i++) {
+                for (int j = 0; j < group_locations.size(); j++) {
+                    if (isVisited[i][j] == 1) continue;
+                    if (Point.getDistance(targets.get(i), group_locations.get(j)) < 0.1) {
+                        isVisited[i][j] = 1;
+                    }
+                }
+            }
             double min = 1e9;
             int mark = 0;
             for (int i = 0; i < targets.size(); i++) {
                 if (visited_set != null && visited_set.get(id).contains(i)) continue;
-                double dist = Point.getDistance(group_locations.get(id), targets.get(i));
+                int count = group_locations.size();
+                for (int j = 0; j < group_locations.size(); j++) {
+                    if (j == id) continue;
+                    if (isVisited[i][j] > 0) {
+                        count--;
+                    }
+                }
+                double dist = getTrueWeight(group_locations.get(id), targets.get(i)) / count;

```

Comment:
How does this version fare when compared to the version where the player goes to the nearest target?
We should try a game with:
Player1 = going to nearest node in a straight line
Player2 = going to nearest node in best path
Player3 = going to "best" node (this version) in best path

And see who performs better to get an idea.

Reply:
Yes I have already tested some cases. Player 2 is usually better than Player 1 because best path is at least not worse than straight line. Player 3 is almost similar to Player 2, maybe a little bit better because he selects the 'true' nearest target. The performance depends on distribution of targets.

Reply:
Interesting. I've seen that the difference in performance is usually prominent when there are more targets, say 10.